### ISSUE TITLE
MoE DDP + Expert Parallelism

### ISSUE THREAD

**Issue by santurini** [laekov/fastmoe#148]
I wanted to know which major changes should I make to a classic DDP torch loop and my architecture in order to make it work.
I would like to change my gated MLPs with a MoE, and given that I have 2 GPUs on 2 different nodes (1 gpu per node), I would like to use 2 experts and have one expert per gpu such that the memory requirements for a single GPU are the same as before.

Is it possible to have something like that? What should I change? Can I use whatever module or are there some constraints?
Also, how does it work for checkpointing, I suppose is the same as DDP right?

Sorry for all this questions but this library is my last hope to put an expert per gpu. I would really appreciate your help!

**Comment by laekov** [comment]
You can use `DistributedGroupedDataParallel` as defined [here](https://github.com/laekov/fastmoe/blob/master/fmoe/distributed.py#LL10C25-L10C25) to replace native DDP of PyTorch, which should have an identical interface.

Also, you need to manually call `model.allreduce_gradients()` after `loss.backward()`.

I suppose the above steps should be enough for your case.

**Comment by laekov** [comment]
For model parameters, it is true that only the extra gating module involves extra memory consumption. However, the saved activations of MoE layers may also consume some extra memory.

**Comment by laekov** [comment]
Yes, I think so.

**Comment by santurini** [comment]
Great! Thank you, I'll let you know.
